File: android/app/tests/unit/src/com/android/bluetooth/le_audio/ContentControlIdKeeperTest.java
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.le_audio;

import static org.mockito.Mockito.*;

import android.bluetooth.BluetoothLeAudio;
import android.os.ParcelUuid;

import android.os.ParcelUuid;
import android.util.Pair;
import androidx.test.InstrumentationRegistry;
import androidx.test.filters.MediumTest;
import androidx.test.runner.AndroidJUnit4;

import com.android.bluetooth.btservice.ServiceFactory;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Map;
import java.util.UUID;

@MediumTest
@RunWith(AndroidJUnit4.class)
public class ContentControlIdKeeperTest {
    @Mock
    ServiceFactory mServiceFactoryMock;
    @Mock
    LeAudioService mLeAudioServiceMock;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        doReturn(mLeAudioServiceMock).when(mServiceFactoryMock).getLeAudioService();
        ContentControlIdKeeper.initForTesting(mServiceFactoryMock);
    }

    @After
    public void tearDown() throws Exception {
        ContentControlIdKeeper.initForTesting(null);
    }

    public int testCcidAcquire(ParcelUuid uuid, int context, int expectedListSize) {
        int ccid = ContentControlIdKeeper.acquireCcid(uuid, context);
        Assert.assertNotEquals(
                ccid,
                ContentControlIdKeeper.CCID_INVALID);

        verify(mLeAudioServiceMock).setCcidInformation(eq(uuid),
                        eq(ccid), eq(context));
        Map<ParcelUuid, Pair<Integer, Integer>> uuidToCcidContextPair =
                ContentControlIdKeeper.getUuidToCcidContextPairMap();
        Assert.assertEquals(expectedListSize, uuidToCcidContextPair.size());
        Assert.assertTrue(uuidToCcidContextPair.containsKey(uuid));
        Assert.assertEquals(ccid, (long)uuidToCcidContextPair.get(uuid).first);
        Assert.assertEquals(context, (long)uuidToCcidContextPair.get(uuid).second);

        return ccid;
    }

    public void testCcidRelease(ParcelUuid uuid, int ccid, int expectedListSize) {
        Map<ParcelUuid, Pair<Integer, Integer>> uuidToCcidContextPair =
                        ContentControlIdKeeper.getUuidToCcidContextPairMap();
        Assert.assertTrue(uuidToCcidContextPair.containsKey(uuid));

        ContentControlIdKeeper.releaseCcid(ccid);
        uuidToCcidContextPair = ContentControlIdKeeper.getUuidToCcidContextPairMap();
        Assert.assertFalse(uuidToCcidContextPair.containsKey(uuid));

        verify(mLeAudioServiceMock).setCcidInformation(eq(uuid),
                eq(ccid), eq(0));

        Assert.assertEquals(expectedListSize, uuidToCcidContextPair.size());
    }

    @Test
    public void testAcquireReleaseCcid() {
        ParcelUuid uuid_one = new ParcelUuid(UUID.randomUUID());
        ParcelUuid uuid_two = new ParcelUuid(UUID.randomUUID());

        int ccid_one = testCcidAcquire(uuid_one, BluetoothLeAudio.CONTEXT_TYPE_MEDIA, 1);
        int ccid_two = testCcidAcquire(uuid_two, BluetoothLeAudio.CONTEXT_TYPE_RINGTONE, 2);
        Assert.assertNotEquals(ccid_one, ccid_two);

        testCcidRelease(uuid_one, ccid_one, 1);
        testCcidRelease(uuid_two, ccid_two, 0);
    }

    @Test
    public void testAcquireReleaseCcidForCompoundContext() {
        ParcelUuid uuid = new ParcelUuid(UUID.randomUUID());
        int ccid = testCcidAcquire(uuid,
                BluetoothLeAudio.CONTEXT_TYPE_MEDIA | BluetoothLeAudio.CONTEXT_TYPE_RINGTONE, 1);
        testCcidRelease(uuid, ccid, 0);
    }

    @Test
    public void testAcquireInvalidContext() {
        ParcelUuid uuid = new ParcelUuid(UUID.randomUUID());

        int ccid = ContentControlIdKeeper.acquireCcid(uuid, 0);
        Assert.assertEquals(ccid, ContentControlIdKeeper.CCID_INVALID);

        verify(mLeAudioServiceMock,
                times(0)).setCcidInformation(any(ParcelUuid.class), any(int.class), any(int.class));
        Map<ParcelUuid, Pair<Integer, Integer>> uuidToCcidContextPair =
                ContentControlIdKeeper.getUuidToCcidContextPairMap();
        Assert.assertEquals(0, uuidToCcidContextPair.size());
    }

    @Test
    public void testAcquireContextMoreThanOnce() {
        ParcelUuid uuid = new ParcelUuid(UUID.randomUUID());

        int ccid_one = testCcidAcquire(uuid, BluetoothLeAudio.CONTEXT_TYPE_MEDIA, 1);
        int ccid_two = testCcidAcquire(uuid, BluetoothLeAudio.CONTEXT_TYPE_RINGTONE, 1);

        // This is implementation specific but verifies that the previous CCID was recycled
        Assert.assertEquals(ccid_one, ccid_two);
    }

}

File: system/bta/le_audio/content_control_id_keeper_test.cc
/******************************************************************************
 *
 * Copyright (c) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************************/

#include "content_control_id_keeper.h"

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include "le_audio_types.h"

namespace le_audio {

TEST(ContentControlIdKeeperTest, testStart) {
  auto ccid_keeper = ContentControlIdKeeper::GetInstance();
  ASSERT_NE(nullptr, ccid_keeper);
  ccid_keeper->Start();
  ccid_keeper->Stop();
}

TEST(ContentControlIdKeeperTest, testMultipleSetGetOneCcid) {
  auto ccid_keeper = ContentControlIdKeeper::GetInstance();
  ASSERT_NE(nullptr, ccid_keeper);
  ccid_keeper->Start();

  int ccid_one = 1;

  ccid_keeper->SetCcid(
      types::LeAudioContextType::MEDIA | types::LeAudioContextType::ALERTS,
      ccid_one);
  ASSERT_EQ(ccid_one, ccid_keeper->GetCcid(types::LeAudioContextType::MEDIA));
  ASSERT_EQ(ccid_one, ccid_keeper->GetCcid(types::LeAudioContextType::ALERTS));

  auto media_ccids = ccid_keeper->GetAllCcids(
      types::AudioContexts(types::LeAudioContextType::MEDIA));
  ASSERT_EQ(1LU, media_ccids.size());
  ASSERT_NE(media_ccids.end(),
            std::find(media_ccids.begin(), media_ccids.end(), ccid_one));

  auto allerts_ccids = ccid_keeper->GetAllCcids(
      types::AudioContexts(types::LeAudioContextType::ALERTS));
  ASSERT_EQ(1LU, allerts_ccids.size());
  ASSERT_NE(allerts_ccids.end(),
            std::find(allerts_ccids.begin(), allerts_ccids.end(), ccid_one));

  auto all_ccids = ccid_keeper->GetAllCcids(types::LeAudioContextType::MEDIA |
                                            types::LeAudioContextType::ALERTS);
  ASSERT_EQ(1LU, all_ccids.size());
  ASSERT_NE(all_ccids.end(),
            std::find(all_ccids.begin(), all_ccids.end(), ccid_one));

  ccid_keeper->Stop();
}

TEST(ContentControlIdKeeperTest, testMultipleSetGetMultipleCcids) {
  auto ccid_keeper = ContentControlIdKeeper::GetInstance();
  ASSERT_NE(nullptr, ccid_keeper);
  ccid_keeper->Start();

  int ccid_two = 2;
  int ccid_three = 3;

  ccid_keeper->SetCcid(types::LeAudioContextType::MEDIA, ccid_two);
  ASSERT_EQ(ccid_two, ccid_keeper->GetCcid(types::LeAudioContextType::MEDIA));
  ccid_keeper->SetCcid(types::LeAudioContextType::ALERTS, ccid_three);
  ASSERT_EQ(ccid_three,
            ccid_keeper->GetCcid(types::LeAudioContextType::ALERTS));

  auto media_ccids = ccid_keeper->GetAllCcids(
      types::AudioContexts(types::LeAudioContextType::MEDIA));
  ASSERT_EQ(1LU, media_ccids.size());
  ASSERT_NE(media_ccids.end(),
            std::find(media_ccids.begin(), media_ccids.end(), ccid_two));

  auto allerts_ccids = ccid_keeper->GetAllCcids(
      types::AudioContexts(types::LeAudioContextType::ALERTS));
  ASSERT_EQ(1LU, allerts_ccids.size());
  ASSERT_NE(allerts_ccids.end(),
            std::find(allerts_ccids.begin(), allerts_ccids.end(), ccid_three));

  auto all_ccids = ccid_keeper->GetAllCcids(types::LeAudioContextType::MEDIA |
                                            types::LeAudioContextType::ALERTS);
  ASSERT_EQ(2LU, all_ccids.size());
  ASSERT_NE(all_ccids.end(),
            std::find(all_ccids.begin(), all_ccids.end(), ccid_two));
  ASSERT_NE(all_ccids.end(),
            std::find(all_ccids.begin(), all_ccids.end(), ccid_three));

  ccid_keeper->Stop();
}

TEST(ContentControlIdKeeperTest, testStop) {
  auto ccid_keeper = ContentControlIdKeeper::GetInstance();
  ASSERT_NE(nullptr, ccid_keeper);
  ccid_keeper->Start();

  int ccid_two = 2;
  int ccid_three = 3;
  ccid_keeper->SetCcid(types::LeAudioContextType::MEDIA, ccid_two);
  ccid_keeper->SetCcid(types::LeAudioContextType::ALERTS, ccid_three);

  ccid_keeper->Stop();

  // Check if it all got erased on Stop
  ASSERT_EQ(-1, ccid_keeper->GetCcid(types::LeAudioContextType::MEDIA));
  ASSERT_EQ(-1, ccid_keeper->GetCcid(types::LeAudioContextType::ALERTS));
  auto all_ccids = ccid_keeper->GetAllCcids(types::LeAudioContextType::MEDIA |
                                            types::LeAudioContextType::ALERTS);
  ASSERT_EQ(0LU, all_ccids.size());
}

}  // namespace le_audio
